### PR TITLE
fixes #26495 - graphql: properly auth resources

### PR DESCRIPTION
**Pull request by timogoebel**

### CODE REVIEW

**Comment by tbrisker** [comment]
test failures related

**Comment by timogoebel** [comment]
Fixed the tests.

File: app/graphql/record_loader.rb [hunk 1]
```diff
@@ -4,7 +4,21 @@ def initialize(model)
...
+    load(id)
+  end
+
+  private
+
+  def authorized_scope
+    return @model unless @model.respond_to?(:authorized)
+
+    permission_name = @model.find_permission_name(:view)
+    @model.authorized_as(User.current, permission_name)
```
Comment (tbrisker) [inline app/graphql/record_loader.rb:22]: Should this use the context current_user? (I'm not really sure I understand why in certain places we pass the context around and others not)

File: app/graphql/types/base_object.rb [hunk 2]
```diff
@@ -36,23 +36,31 @@ def belongs_to(name, type, resolver: nil)
...
         reflection = model_class.reflect_on_association(attribute)
-        return false if reflection && reflection.macro == :belongs_to && nullable?(reflection.foreign_key)
+        return true if reflection && reflection.macro == :belongs_to && attribute_required?(reflection.foreign_key)
 
-        true
+        false
+      end
+
+      def nullable?(attribute)
+        !attribute_required?(attribute)
```
Comment (tbrisker) [inline app/graphql/types/base_object.rb:63]: is this change related? I don't see anywhere else that calls these functions but i may be missing something. If it's unrelated please split it off to a separate pr

File: app/graphql/collection_loader.rb [hunk 3]
```diff
@@ -2,7 +2,7 @@ class CollectionLoader < GraphQL::Batch::Loader
   def initialize(model, association_name, scope = nil)
     @model = model
     @association_name = association_name
-    @scope = scope
+    @scope = scope || authorized_scope
```
Comment (tbrisker) [inline app/graphql/collection_loader.rb]: This means if a consumer of this class passes a scope it could bypass the authorization by mistake, e.g. in Resolvers::Domain::Subnets.

**Review by tbrisker — CHANGES_REQUESTED** [review]

File: app/graphql/types/base_object.rb (hunk 2 above)
Comment (timogoebel) [reply app/graphql/types/base_object.rb:63]: It's not directly related, but tests won't pass with out this. So we need it before we can merge this PR.

File: app/graphql/record_loader.rb (hunk 1 above)
Comment (timogoebel) [reply app/graphql/record_loader.rb:22]: GraphQL expected that the current user is passed via the context, but we rely on `User.current` for writing Audits etc. So we basically need both.
In the future, we should probably unify this, but I think right now we can live with providing both options.

File: app/graphql/collection_loader.rb (hunk 3 above)
Comment (timogoebel) [reply app/graphql/collection_loader.rb]: Ok, this was a bad shortcut I took. A lambda that just extends the `authorized_scope` is a much better idea.

**Comment by timogoebel** [comment]
@tbrisker: Thanks for your comments. Love your reviews (I really do, good job!). I hope it's ok if I ignore two of your comments, for now, as we can work on them later.
The ability to override the authorized scope was a really bad judgement call on my end, I've replaced this with the ability to pass a lambda that just extends the base scope.

In a follow-up PR, I completely want to get rid of `Queries::AuthorizedModelQuery` and refactor the functionality in a resolver. That resolver can then also handle filtering by `Organization` and `Location`.

**Review by tbrisker — APPROVED** [review]
Thanks for the kind words and the fix @timogoebel !
